### PR TITLE
Add support for dimensions for gauges

### PR DESCRIPTION
**Pull request by rlinehan**
Add support for dimensioned gauges. Dimensions are added to a gauge when it is created. Gauges are stored in a `GaugeGroup` by their `DimensionPosition`.

Dimensioned gauges are created by building up a `GaugeDimensions`, which is a group of dimensions, and then providing that when creating the gauge. The `GaugeDimensions` is consumed on gauge creation.

Example:
```rust
metrics_factory.dimensioned_gauge("gauge_group", "gauge", GaugeDimensions::from("a", "dimension"))
```

### CODE REVIEW

File: lib/src/gauge.rs [hunk 1]
```diff
@@ -72,3 +76,66 @@ impl StatisticSetGauge {
         })
     }
 }
+
+/// A GaugeDimensions is a group of dimensions to be used with a gauge.
+/// It is consumed when the gauge is created.
+#[derive(Debug, Default)]
+pub struct GaugeDimensions {
+    pub dimension_position: DimensionPosition,
```
Comment (kvc0) [inline lib/src/gauge.rs:84]: Why is this public? It's not useful to library consumers. Maybe you want to `impl From<GaugeDimensions> for DimensionPosition` instead?

File: lib/src/gauge.rs [hunk 2]
```diff
@@ -72,3 +76,66 @@ impl StatisticSetGauge {
...
+}
+
+impl GaugeDimensions {
+    /// Create a new, empty, GaugeDimensions
+    pub fn new() -> Self {
+        Self {
+            dimension_position: Default::default(),
+        }
+    }
```
Comment (kvc0) [inline lib/src/gauge.rs:93]: Is there a reason why this type should not have a trimmer api? E.g.,

```rust
/// Create a new GaugeDimensions
/// ```
/// # use goodmetrics::gauge::GaugeDimensions;
/// GaugeDimensions::new([("a", "dimension"), ("another", "dimension")]);
/// ```
pub fn new(dimensions: impl IntoIterator<Item = (impl Into<Name>, impl Into<Dimension>)>) -> Self {
    Self {
        dimension_position: BTreeMap::from_iter(dimensions.into_iter().map(|(name, dimension)| (name.into(), dimension.into()))),
    }
}
```

and no "from" "with". You already have a Default impl derived, so there's no value to this `new()` as-is. Default + insert accounts for people who want to go one-by-one, and new() in this way accounts for people who want to express their dimensions declaratively.

File: lib/src/gauge.rs [hunk 3]
```diff
@@ -72,3 +76,66 @@ impl StatisticSetGauge {
...
+    /// ```
+    /// use goodmetrics::gauge::GaugeDimensions;
+    ///
+    /// GaugeDimensions::from("a", "dimension");
+    /// ```
+    pub fn from(name: impl Into<Name>, dimension: impl Into<Dimension>) -> Self {
+        Self {
+            dimension_position: BTreeMap::from([(name.into(), dimension.into())]),
+        }
+    }
```
Comment (kvc0) [inline lib/src/gauge.rs:106]: This is a copy of insert and with_dimension, semantically. It's a rename of the api in the case of the first element. I think this, as well as with_dimension should be removed.

File: lib/src/gauge_group.rs [hunk 4]
```diff
@@ -3,46 +3,76 @@ use std::{
...
     gauge::{self, StatisticSetGauge},
-    pipeline::aggregation::statistic_set::StatisticSet,
+    pipeline::aggregator::DimensionPosition,
     types::Name,
 };
 
 #[derive(Default)]
 pub struct GaugeGroup {
-    gauges: HashMap<Name, Weak<StatisticSetGauge>>,
+    pub(crate) dimensioned_gauges:
```
Comment (kvc0) [inline lib/src/gauge_group.rs:16]: Why is this pub(crate)?

File: lib/src/gauge_group.rs [hunk 5]
```diff
@@ -3,46 +3,76 @@ use std::{
...
 
-        self.gauges.iter().filter_map(|(name, possible_gauge)| {
-            possible_gauge
-                .upgrade()
-                .and_then(|gauge| gauge.reset())
-                .map(|statistic_set| (name.to_owned(), statistic_set))
-        })
+        self.dimensioned_gauges
+            .iter()
+            .map(|(dimension_position, gauges)| {
```
Comment (kvc0) [inline lib/src/gauge_group.rs:60]: how is this removing dimension positions that are no longer populated with any gauges once you filter_map them out?

File: lib/src/gauge_group.rs [hunk 6]
```diff
@@ -3,46 +3,76 @@ use std::{
...
-                self.gauges.insert(name, Arc::downgrade(&gauge));
+                gauge_position.insert(name, Arc::downgrade(&gauge));
                 gauge
             }
         }
     }
 
-    pub fn reset(&mut self) -> impl Iterator<Item = (Name, StatisticSet)> + '_ {
-        self.gauges.retain(|_name, gauge| gauge.upgrade().is_some());
+    // reset gauge group, aggregating and returning previous values
```
Comment (kvc0) [inline lib/src/gauge_group.rs:50]:
```suggestion
    /// reset gauge group to 0's, returning the current aggregations
```
This should be a `///` not a `//`

File: lib/src/metrics_factory.rs [hunk 7]
```diff
@@ -207,17 +204,28 @@ impl<TMetricsAllocator, TSink> MetricsFactory<TMetricsAllocator, TSink> {
...
+        gauge_dimensions: GaugeDimensions,
     ) -> Arc<StatisticSetGauge> {
         let mut locked_groups = self
             .gauge_groups
             .lock()
             .expect("local mutex should not be poisoned");
         let gauge_group = gauge_group.into();
         match locked_groups.get_mut(&gauge_group) {
-            Some(group) => group.gauge(gauge_name),
+            Some(group) => group.dimensioned_gauge(gauge_name, gauge_dimensions.dimension_position),
```
Comment (kvc0) [inline lib/src/metrics_factory.rs:224]: this should be `gauge_dimensions.into()`

**Review by kvc0 — COMMENTED** [review]
Thanks for the feature. I've got a couple minor notes

File: lib/src/gauge.rs (hunk 2 above)
Comment (rlinehan) [reply lib/src/gauge.rs:93]: That's much better! Thanks.

File: lib/src/gauge_group.rs (hunk 4 above)
Comment (rlinehan) [reply lib/src/gauge_group.rs:16]: I think just a leftover from some earlier code I had, I'll remove.

File: lib/src/gauge_group.rs (hunk 5 above)
Comment (rlinehan) [reply lib/src/gauge_group.rs]: good call, fixed in [c6a833a](https://github.com/kvc0/goodmetrics_rs/pull/13/commits/c6a833a801d30f08f0bf673ac3e1dcf4d1478825) and also added a test

File: lib/src/metrics_factory.rs [hunk 8]
```diff
@@ -410,4 +420,120 @@ mod test {
 
         let _metrics_that_shares_the_sink = cloned.record_scope("scope_name");
     }
+
+    #[test_log::test(tokio::test)]
+    async fn gauges() {
```
Comment (kvc0) [inline lib/src/metrics_factory.rs:425]: great test

**Review by kvc0 — APPROVED** [review]
This looks good. Thanks again for the feature.

**Pull request merged**